### ISSUE TITLE
IllegalArgumentException: Unable to resolve classname: IPersistentMap

### ISSUE THREAD

**Issue by nicksieger** [amalloy/ordered#24]
Here's a failing testcase for `map_test.clj`:

```clojure
(deftest read-direct
  (let [s (ordered-map 1 2, 3 4, 5 6, 1 9, 7 8)
        t #ordered/map ([1 9] [3 4] [5 6] [7 8])]
    (is (= s t))))
```

This blows up with the error message in the issue title.

If I change code in `map.clj` to fully-qualify all instances of `IPersistentMap`, then the error message changes to "java.lang.IllegalArgumentException: No matching field found: backing-map for class flatland.ordered.map.OrderedMap".

My clojure-fu is not skillful enough to unwind your `delegating-deftype` macro, so I thought I'd ask here and see if you know what's going on.

Thanks!

**Comment by amalloy** [comment]
You're not really supposed to use data readers to embed foreign objects in your source code. Most of your test case is just distracting fluff: it's easily reproduced with the simple expression 

```clj
    #ordered/map ([1 2])
```

I don't know the details of how the compiler tries to understand map objects embedded in code, which make it come up with this exception, but basically it is just something you're not supposed to do. The tagged syntax is for transporting stuff across the wire, or persisting it to disk. That is, 

```clj
    (read-string "#ordered/map ([1 2])")
```

works fine, and that's what the feature is supposed to do. What you're doing is the same as 

```clj
    (eval (read-string "#ordered/map ([1 2])"))
```

which fails for a lot of reader-tagged objects. Do you have some compelling reason that you need this to work? I have trouble imagining a way in which the answer could be yes.

**Comment by nicksieger** [comment]
I understand now that use of reader-tagged objects in bare code is not something that you want to do typically. I encountered this error in code that embedded an ordered map in metadata while passing through a couple layers of macros. Switching away from an ordered map to a regular map made the issue go away. My best guess after finding out that I could trigger the error with `#ordered/map ([1 2])` was that somewhere through the layers of macros the ordered map was being rendered by `print-method`. I didn't find the exact culprit unfortunately (dense code), so it was still a guess at best.

I was able to fix this issue by fully-qualifying interfaces and protocols in `map.clj` and changing the `backing-map` field name to `backing_map` (to avoid getting munged by the compiler). (Why fully-qualifying was necessary is beyond my comprehension of clojure's evaluator.) If you are interested in those changes I can put up a PR, but at any rate I agree with you that I don't think this is an issue that needs to be solved in your library.
